File: python/combined/camera_lidar.py
# SPDX-License-Identifier: Apache-2.0
# Copyright © 2025 Au-Zone Technologies. All Rights Reserved.

from argparse import ArgumentParser
import asyncio
import io
import sys
import av
import zenoh
import threading
import time
from collections import deque
import numpy as np
import rerun as rr
import rerun.blueprint as rrb
from edgefirst.schemas.edgefirst_msgs import Detect
from edgefirst.schemas.sensor_msgs import PointCloud2
from edgefirst.schemas import decode_pcd, colormap, turbo_colormap
import threading

def format_remote_endpoint(remote):
    if not remote:
        return None
    
    # Already in full format (tcp/IP:PORT)
    if remote.startswith("tcp/") and ":" in remote.split("/", 1)[1]:
        return remote
    
    # Just IP address (e.g., "10.10.41.100")
    if "/" not in remote and ":" not in remote:
        return f"tcp/{remote}:7447"
    
    # IP:PORT format (e.g., "10.10.41.100:7447")
    if ":" in remote and not remote.startswith("tcp/"):
        return f"tcp/{remote}"
    
    # tcp/IP format (e.g., "tcp/10.10.41.100")
    if remote.startswith("tcp/") and ":" not in remote.split("/", 1)[1]:
        return f"{remote}:7447"
    
    # Fallback: return as-is
    return remote

# Global variables for message storage
h264_msg = None
boxes2d_msg = None
lidar_msg = None

# Global for frame size tracking
frame_size = FrameSize()


class FrameSize:
    def __init__(self):
        self._size = []
        self._event = asyncio.Event()

    def set(self, width, height):
        self._size = [width, height]
        if not self._event.is_set():
            self._event.set()

    async def get(self):
        await self._event.wait()
        return self._size


def h264_handler(msg):
    """Simple sync handler that stores message in global."""
    global h264_msg
    h264_msg = msg


def boxes2d_handler(msg):
    """Simple sync handler that stores message in global."""
    global boxes2d_msg
    boxes2d_msg = msg


def clusters_handler(msg):
    """Simple sync handler that stores message in global."""
    global lidar_msg
    lidar_msg = msg


async def h264_worker():
    """Async worker that processes H.264 messages from global."""
    global h264_msg, frame_size
    raw_data = io.BytesIO()
    container = av.open(raw_data, format="h264", mode="r")
    
    while True:
        if h264_msg is not None:
            try:
                msg = h264_msg
                raw_data.write(msg.payload.to_bytes())
                raw_data.seek(0)
                for packet in container.demux():
                    try:
                        if packet.size == 0:
                            continue
                        raw_data.seek(0)
                        raw_data.truncate(0)
                        for frame in packet.decode():
                            frame_array = frame.to_ndarray(format="rgb24")
                            frame_size.set(frame_array.shape[1], frame_array.shape[0])
                            rr.log("/camera", rr.Image(frame_array))
                    except Exception:
                        continue
            except Exception as e:
                print(f"Error processing H.264 message: {e}", file=sys.stderr)
        await asyncio.sleep(0.01)


async def boxes2d_worker():
    """Async worker that processes boxes2d messages from global."""
    global boxes2d_msg, frame_size
    boxes_tracked = {}
    _ = await frame_size.get()
    
    while True:
        if boxes2d_msg is not None:
            try:
                msg = boxes2d_msg
                detection = Detect.from_cdr(msg.payload.to_bytes())
                current_frame_size = await frame_size.get()
                centers, sizes, labels, colors = [], [], [], []
                for box in detection.boxes:
                    if box.track.id and box.track.id not in boxes_tracked:
                        boxes_tracked[box.track.id] = [
                            box.label + ": " + box.track.id[:6],
                            list(np.random.choice(range(256), size=3)),
                        ]
                    if box.track.id:
                        colors.append(boxes_tracked[box.track.id][1])
                        labels.append(boxes_tracked[box.track.id][0])
                    else:
                        colors.append([0, 255, 0])
                        labels.append(box.label)
                    centers.append(
                        (int(box.center_x * current_frame_size[0]), int(box.center_y * current_frame_size[1]))
                    )
                    sizes.append((int(box.width * current_frame_size[0]), int(box.height * current_frame_size[1])))
                rr.log(
                    "/camera/boxes",
                    rr.Boxes2D(centers=centers, sizes=sizes, labels=labels, colors=colors),
                )
            except Exception as e:
                print(f"Error processing boxes2d message: {e}", file=sys.stderr)
        await asyncio.sleep(0.01)


async def lidar_worker():
    """Async worker that processes LiDAR cluster messages from global."""
    global lidar_msg
    
    while True:
        if lidar_msg is not None:
            try:
                msg = lidar_msg
                pcd = PointCloud2.from_cdr(msg.payload.to_bytes())
                points = decode_pcd(pcd)
                clusters = [p for p in points if p.cluster_id > 0]
                if not clusters:
                    rr.log("/pointcloud/clusters", rr.Points3D([], colors=[]))
                else:
                    max_id = max(p.cluster_id for p in clusters)
                    pos = [[p.x, p.y, p.z] for p in clusters]
                    colors = [colormap(turbo_colormap, p.cluster_id / max_id) for p in clusters]
                    rr.log("/pointcloud/clusters", rr.Points3D(pos, colors=colors))
            except Exception as e:
                print(f"Error processing LiDAR cluster message: {e}", file=sys.stderr)
        await asyncio.sleep(0.01)


async def main_async(args):
    # Setup rerun
    # args.memory_limit = 10
    rr.script_setup(args, "camera-lidar")

    blueprint = rrb.Blueprint(
        rrb.Grid(
            contents=[
                rrb.Spatial2DView(origin="/camera", name="Camera Feed"),
                rrb.Spatial3DView(origin="/pointcloud", name="Pointcloud Clusters"),
            ]
        )
    )
    rr.send_blueprint(blueprint)

    # Zenoh config
    config = zenoh.Config()
    config.insert_json5("scouting/multicast/interface", "'lo'")
    remote = format_remote_endpoint(args.remote)
    if remote:
        config.insert_json5("mode", "'client'")
        config.insert_json5("connect", f'{{"endpoints": ["{remote}"]}}')
    session = zenoh.open(config)

    # Declare subscribers with simple handlers
    session.declare_subscriber("rt/camera/h264", h264_handler)
    session.declare_subscriber("rt/model/boxes2d", boxes2d_handler)
    session.declare_subscriber("rt/lidar/clusters", clusters_handler)

    # Launch concurrent processing tasks
    try:
        await asyncio.gather(
            h264_worker(),
            boxes2d_worker(),
            lidar_worker(),
        )
    except KeyboardInterrupt:
        pass


def main():
    parser = ArgumentParser(description="EdgeFirst Samples - Camera-Lidar")
    parser.add_argument(
        "-r",
        "--remote",
        type=str,
        default=None,
        help="Remote endpoint (IP:PORT or just IP, defaults to port 7447)",
    )
    rr.script_add_args(parser)
    args = parser.parse_args()

    try:
        asyncio.run(main_async(args))
    except KeyboardInterrupt:
        sys.exit(0)


if __name__ == "__main__":
    main()

File: rust/combined/camera_lidar.rs
use std::{
    error::Error,
    sync::Arc
};
use clap::Parser;
use edgefirst_samples::Args;
use edgefirst_schemas::{
    decode_pcd,
    edgefirst_msgs::Detect,
    foxglove_msgs::FoxgloveCompressedVideo,
    sensor_msgs::{PointCloud2},
};
use openh264::{decoder::Decoder, formats::YUVSource, nal_units};
use rerun::{Boxes3D, Color, Image, Points3D, Position3D};
use tokio::{sync::Mutex, task};
use zenoh::{handlers::FifoChannelHandler, pubsub::Subscriber, sample::Sample};

async fn camera_h264_handler(
    sub: Subscriber<FifoChannelHandler<Sample>>,
    rr: Arc<Mutex<rerun::RecordingStream>>,
    frame_size: Arc<Mutex<[u32; 2]>>,
) {
    // Create decoder inside the function
    let mut decoder = Decoder::new().expect("Failed to create decoder");

    while let Ok(msg) = sub.recv_async().await {
        let video = match cdr::deserialize::<FoxgloveCompressedVideo>(&msg.payload().to_bytes()) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Failed to deserialize video: {:?}", e);
                continue;
            }
        };

        for packet in nal_units(&video.data) {
            let Ok(Some(yuv)) = decoder.decode(packet) else { continue };
            let rgb_len = yuv.rgb8_len();
            let mut rgb_raw = vec![0; rgb_len];
            yuv.write_rgb8(&mut rgb_raw);
            let width = yuv.dimensions().0;
            let height = yuv.dimensions().1;
            let image = Image::from_rgb24(rgb_raw, [width as u32, height as u32]);
            let rr_guard = rr.lock().await;
            if let Err(e) = rr_guard.log("/camera", &image) {
                eprintln!("Failed to log video: {:?}", e);
            }

            let mut frame_size = frame_size.lock().await;
            *frame_size = [width as u32, height as u32];
        }
    }
}

async fn model_boxes2d_handler(
    sub: Subscriber<FifoChannelHandler<Sample>>,
    rr: Arc<Mutex<rerun::RecordingStream>>,
    frame_size: Arc<Mutex<[u32; 2]>>,
) {
    while let Ok(msg) = sub.recv_async().await {
        let detection = match cdr::deserialize::<Detect>(&msg.payload().to_bytes()) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Failed to deserialize detect message: {:?}", e);
                continue; // skip this message and continue
            }
        };
        let mut centers = Vec::new();
        let mut sizes = Vec::new();
        let mut labels = Vec::new();
        let size = frame_size.lock().await;
        if size[0] == 0 || size[1] == 0 { continue; }

        for b in detection.boxes {
            centers.push([b.center_x * size[0] as f32, b.center_y * size[1] as f32]);
            sizes.push([b.width * size[0] as f32, b.height * size[1] as f32]);
            labels.push(b.label);
        }
        drop(size);

        let rr_guard = rr.lock().await;
        let _ = match rr_guard.log("/camera/boxes2d", &rerun::Boxes2D::from_centers_and_sizes(centers, sizes).with_labels(labels)) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("Failed to log boxes2d: {:?}", e);
                continue; // skip this message and continue
                }
            };   
    }
}

async fn lidar_clusters_handler(
    sub: Subscriber<FifoChannelHandler<Sample>>,
    rr: Arc<Mutex<rerun::RecordingStream>>,
) {
    while let Ok(msg) = sub.recv_async().await {
        let pcd = match cdr::deserialize::<PointCloud2>(&msg.payload().to_bytes()) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Failed to deserialize radar pointcloud: {:?}", e);
                continue; // skip this message and continue
            }
        };
        let points = decode_pcd(&pcd);
        let clustered_points: Vec<_> = points.iter().filter(|x| x.id > 0).collect();
        let max_cluster_id = clustered_points
            .iter()
            .map(|x| x.id)
            .max()
            .unwrap_or(1)
            .max(1);

        let points = Points3D::new(
            clustered_points
                .iter()
                .map(|p| Position3D::new(p.x as f32, p.y as f32, p.z as f32)),
        )
        .with_colors(clustered_points.iter().map(|p| {
            let (r, g, b) = colorous::TURBO
                .eval_continuous(p.id as f64 / max_cluster_id as f64)
                .as_tuple();
            Color::from_rgb(r, g, b)
        }));
        let rr_guard = rr.lock().await;
        let _ = match rr_guard.log("/pointcloud/lidar", &points) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("Failed to log radar pointcloud: {:?}", e);
                continue; // skip this message and continue
                }
            };  
    }
}

async fn fusion_boxes3d_handler(
    sub: Subscriber<FifoChannelHandler<Sample>>,
    rr: Arc<Mutex<rerun::RecordingStream>>,
) {
    while let Ok(msg) = sub.recv_async().await {
        let det = match cdr::deserialize::<Detect>(&msg.payload().to_bytes()) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Failed to deserialize fusion_boxes3d: {:?}", e);
                continue; // skip this message and continue
            }
        };
        let boxes = det.boxes;
        // The 3D boxes are in an _optical frame of reference, where x is right, y is down, and z (distance) is forward
        // We will convert them to a normal frame of reference, where x is forward, y is left, and z is up
        let rr_boxes = Boxes3D::from_centers_and_sizes(
            boxes.iter().map(|b| (b.distance, -b.center_x, -b.center_y)),
            boxes.iter().map(|b| (b.width, b.width, b.height)),
        );
        let rr_guard = rr.lock().await;
        let _ = match rr_guard.log("/pointcloud/boxes3d", &rr_boxes) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Failed to log fusion boxes3d: {:?}", e);
                continue; // skip this message and continue
            }
        };
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args = Args::parse();
    let session = zenoh::open(args.clone()).await.unwrap();

    let (rr, _serve_guard) = args.rerun.init("camera-lidar")?;
    let rr = Arc::new(Mutex::new(rr));
    let frame_size = Arc::new(Mutex::new([0u32; 2]));

    let cam_sub = session.declare_subscriber("rt/camera/h264").await.unwrap();
    let cam_rr = rr.clone();
    let frame_size_cam = frame_size.clone();
    task::spawn(camera_h264_handler(cam_sub, cam_rr, frame_size_cam));

    let boxes2d_sub = session.declare_subscriber("rt/model/boxes2d").await.unwrap();
    let boxes2d_rr = rr.clone();
    let frame_size_boxes2d = frame_size.clone();
    task::spawn(model_boxes2d_handler(boxes2d_sub, boxes2d_rr, frame_size_boxes2d));

    let lidar_sub = session.declare_subscriber("rt/lidar/clusters").await.unwrap();
    let lidar_rr = rr.clone();
    task::spawn(lidar_clusters_handler(lidar_sub, lidar_rr));

    let boxes3d_sub = session.declare_subscriber("rt/fusion/boxes3d").await.unwrap();
    let boxes3d_rr = rr.clone();
    task::spawn(fusion_boxes3d_handler(boxes3d_sub, boxes3d_rr));

    loop {
    }
}
